Update account docs (#709)
OpenZeppelin/cairo-contracts: docs/modules/ROOT/pages/guides/deployment.adoc

:foundry: https://foundry-rs.github.io/starknet-foundry/starknet/account.html[Starknet Foundry]
:starkli: https://book.starkli.rs/accounts#account-deployment[Starkli]

= Counterfactual deployments

A deployment is said to be counterfactual when the one paying for the deployment fees is the address of
the contract to be deployed. In other words, we need to send the funds to the address even when the
deploymenthasn’t happened.

This process can be described with the following steps:

TIP: For testing this flow you can check the {foundry} or the {starkli} guides for deploying accounts.

1. Deterministically precompute the `contract_address` given a `class_hash`, `salt`, and constructor `calldata`.
Note that the `class_hash` must be previously declared for the deployment to succeed.

2. Send funds to the `contract_address`. Usually you will estimate the fee of the transaction first. Existing
tools usually do this for you.

3. Send a `DeployAccount` type transaction to the network.

> 4. The protocol will then validate the transaction with the `\\__validate_deploy__` entrypoint.

Reviewer: ```suggestion
4. The protocol will then validate the transaction with the `\\__validate_deploy__` entrypoint of the contract to be deployed.
```

5. If the validation succeeds, the protocol will charge the fee and then register the contract as deployed.

> 
> 

Reviewer: ```suggestion

NOTE: Although this method is very popular to deploy accounts, this works for any kind of contract.

```

Reviewer: what do you think about moving this to the beginning of the section?

== Deployment validation

> For counterfactual deployments, contracts must implement the `\\__validate_deploy__` entrypoint, and it is
> called by the protocol when a `DeployAccount` transaction is sent to the network.

Reviewer: ```suggestion
To counterfactually deploy contracts, they must implement the `\\__validate_deploy__` entrypoint,
called by the protocol when a `DeployAccount` transaction is sent to the network.
```

Reviewer: what do you think about moving this to the beginning of the section?

We can leverage this property to make a contract pay for its own deployment by simply sending funds in advance. We call this a counterfactual deployment.

This process can be described with the following steps:

TIP: For testing this flow you can check the {foundry} or the {starkli} guides for deploying accounts.

1. Deterministically precompute the `contract_address` given a `class_hash`, `salt`, and constructor `calldata`.
Note that the `class_hash` must be previously declared for the deployment to succeed.

2. Send funds to the `contract_address`. Usually you will estimate the fee of the transaction first. Existing
tools usually do this for you.

3. Send a `DeployAccount` type transaction to the network.

4. The protocol will then validate the transaction with the `\\__validate_deploy__` entrypoint of the contract to be deployed.

5. If the validation succeeds, the protocol will charge the fee and then register the contract as deployed.

NOTE: Although this method is very popular to deploy accounts, this works for any kind of contract.

== Deployment validation

> To counterfactually deploy contracts, they must implement the `\\__validate_deploy__` entrypoint,

Reviewer 2: ```suggestion
To counterfactually deploy contracts, the target contract must implement the `\\__validate_deploy__` entrypoint,
```
Or maybe "the deploying contract must..." might be better? `they` seems a little vague

Reviewer: "the deploying contract" sounds good
